com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/topology.py

```diff
@@ -2079,19 +2082,49 @@ def _evict_time(self, duration):
         self._config['evictConfig'] = int(duration.total_seconds() * 1000.0)
         self._config['evictTimeUnit'] = 'MILLISECONDS'
 
-    def _partition(self, attribute):
+    def _partition_by_attribute(self, attribute):
         # We cannot always get the list of tuple attributes here
         # because it might be a named type.  Validation of the attribute
         # will be done in code generation.  We only support partition
         # by attribute for StreamSchema (not CommonSchema).
         # Our input schema is the output schema of the previous operator.
         if not isinstance(self.stream.oport.schema, streamsx.topology.schema.StreamSchema):
             raise ValueError("Partition by attribute is supported only for a structured schema")
+
         self._config['partitioned'] = True
         self._config['partitionBy'] = attribute
 
+    def _partition_by_callable(self, function):
+        dilled_callable = None
+
+        # This is based on graph._addOperatorFunction.
+        recurse = None
+        if isinstance(function, types.LambdaType) and function.__name__ == "<lambda>" :
+            function = streamsx.topology.runtime._Callable(function, no_context=True)
+            recurse = True
+        elif function.__module__ == '__main__':
+            # Function/Class defined in main, create a callable wrapping its
+            # dill'ed form
+            function = streamsx.topology.runtime._Callable(function,
+                no_context = True if inspect.isroutine(function) else None)
+            recurse = True
+         
+        if inspect.isroutine(function):
+            # callable is a function
+            name = function.__name__
+        else:
+            # callable is a callable class instance
+            name = function.__class__.__name__
+            # dill format is binary; base64 encode so it is json serializable 
+            dilled_callable = base64.b64encode(dill.dumps(function, recurse=recurse)).decode("ascii")
+
+        self._config['partitioned'] = True
+        if dilled_callable is not None:
+            self._config['partitionByCallable'] = dilled_callable
+        self._config['partitionByName'] = name
+        self._config['partitionByModule'] = function.__module__
 
-    def partition(self, attribute):
+    def partition(self, key):
```

Comment:
Docstring has not been updated for this change, lines 2135, 2140, 2143.

Can you also add a `versionadded` to the docstring with version 1.13.